lib/phoenix/socket.ex

```diff
@@ -47,6 +117,135 @@ defmodule Phoenix.Socket do
             topic: nil,
             transport: nil,
             transport_pid: nil
+
+
+  defmacro __using__(_) do
+    quote do
+      @behaviour Phoenix.Socket
+      import unquote(__MODULE__)
+      Module.register_attribute(__MODULE__, :phoenix_channels, accumulate: true)
+      @phoenix_transports %{}
+
+      transport :websocket, Phoenix.Transports.WebSocket,
+        serializer: Phoenix.Transports.JSONSerializer,
+        timeout: :infinity
+
+      transport :longpoll, Phoenix.Transports.LongPoll,
+        window_ms: 10_000,
+        pubsub_timeout_ms: 1000,
+        crypto: [iterations: 1000, length: 32, digest: :sha256, cache: Plug.Keys]
+
+
+      @before_compile unquote(__MODULE__)
+    end
+  end
+
+  defmacro __before_compile__(env) do
+    transports = Module.get_attribute(env.module, :phoenix_transports)
+    channel_defs =
+      env.module
+      |> Module.get_attribute(:phoenix_channels)
+      |> Helpers.defchannels(transports)
+
+    transport_defs =
+      for {name, {mod, conf}} <- transports do
+        quote do
+          def __transport__(name) when name in [unquote(name), unquote(to_string(name))] do
+            {unquote(mod), unquote(conf)}
+          end
+        end
+      end
+
+    quote do
+      def __transports__, do: unquote(Macro.escape(transports))
+      unquote(transport_defs)
+      def __transport__(_name), do: :unsupported
+      unquote(channel_defs)
+    end
+  end
+
+  @doc """
+  Adds key/value pair to socket assigns.
+
+  ## Examples
+
+      iex> socket.assigns[:token]
+      nil
+      iex> socket = assign(socket, :token, "bar")
+      iex> socket.assigns[:token]
+      "bar"
+
+  """
+  def assign(socket = %Socket{}, key, value) do
+    update_in socket.assigns, &Map.put(&1, key, value)
+  end
+
+  @doc """
+  Defines a channel matching the given topic and transports.
+
+    * `topic_pattern` - The string pattern, ie "rooms:*", "users:*", "system"
+    * `module` - The channel module handler, ie `MyApp.RoomChannel`
+    * `opts` - The optional list of options, see below
+
+  ## Options
+
+    * `:via` - the transport adapters to accept on this channel.
+      Defaults `[:websocket, :longpoll]`
+
+  ## Examples
+
+      channel "topic1:*", MyChannel
+      channel "topic2:*", MyChannel, via: [:websocket]
+      channel "topic",    MyChannel, via: [:longpoll]
+
+  ## Topic Patterns
+
+  The `channel` macro accepts topic patterns in two flavors. A splat argument
+  can be provided as the last character to indicate a "topic:subtopic" match. If
+  a plain string is provied, only that topic will match the channel handler.
+  Most use-cases will use the "topic:*" pattern to allow more versatile topic
+  scoping.
+
+  See `Phoenix.Channel` for more information
+  """
+  defmacro channel(topic_pattern, module, opts \\ []) do
+    quote do
+      @phoenix_channels {unquote_splicing([
+        topic_pattern,
```

Comment:
Why do you need splicing? Is this just a shortcut to avoid unquoting one by one?